<?xml version="1.0"?>
<robot name="test_robot" xmlns:xacro="http://ros.org/wiki/xacro">
  <!-- Define robot constants -->
      <!-- 직선 부분 (edge face) -->
  <xacro:property name="base_f_width" value="0.33"/> 
  <xacro:property name="base_f_length" value="0.33"/>
  <xacro:property name="base_f_height" value="0.325"/>

      <!-- 모서리 부분 (corner element) -->
  <xacro:property name="base_e_width" value="0.2263"/> 
  <xacro:property name="base_e_length" value="0.2263"/>
  <xacro:property name="base_e_height" value="0.325"/>
    
      <!-- lidar 얹을 박스 크기 -->
  <xacro:property name="base_height" value="0.325"/>
  <xacro:property name="box_width" value="0.094"/>
  <xacro:property name="box_length" value="0.104"/>
  <xacro:property name="box_height" value="0.061"/>
  
        <!-- 은판 크기 -->
  <xacro:property name="silver_width" value="0.62"/>
  <xacro:property name="silver_length" value="0.62"/>
  <xacro:property name="silver_height" value="0.006"/>

      <!-- shape of the robot's two wheels -->
  <xacro:property name="wheel_radius" value="0.068"/>
  <xacro:property name="wheel_width" value="0.025"/>
  <xacro:property name="wheel_mass" value="0.5"/>
  
     <!-- y축을 따라 바퀴와 차체 사이의 간격 조정 -->
  <xacro:property name="wheel_yinset" value="0.345"/>
  <!-- <xacro:property name="wheel_yinset" value="${base_width / 2 - wheel_width / 2 - 0.005}"/> -->
     <!-- 각각 z축과 x축을 따라 바퀴의 위치를 조정 -->
  <xacro:property name="wheel_zoff" value="0.13"/>
  <xacro:property name="wheel_xoff" value="0.12"/>
 
         <!-- 캐스터 휠 설정 -->
  <xacro:property name="caster_radius" value="0.05"/>
  <xacro:property name="caster_width" value="0.05"/>
  <xacro:property name="caster_ygap" value="0.24"/>
  <xacro:property name="caster_zoff" value="0.155"/>
  <xacro:property name="caster_xoff_front" value="0.25"/>
  <xacro:property name="caster_xoff_back" value="-0.25"/>
  <xacro:property name="caster_mass" value = "0.3"/>
 
  <!-- 초음파 센서 설정 -->
      <!-- 초음파 센서 전체 크기 -->
  <xacro:property name="ultra_width" value="0.045"/>     <!-- 45 mm -->
  <xacro:property name="ultra_depth" value="0.020"/>      <!-- 20 mm -->
  <xacro:property name="ultra_height" value="0.015"/>     <!-- 15 mm -->

    <!-- 실린더 형태 센서부 -->
  <xacro:property name="ultra_cylinder_radius" value="0.0085"/>
  <xacro:property name="ultra_cylinder_length" value="0.015"/> 

    <!-- 질량 및 관성은 간단화 -->
  <xacro:property name="ultra_mass" value="0.01"/>

  
     <!-- Define inertial property macros  -->
  <xacro:macro name="box_inertia" params="m w h d">
    <inertial>
      <origin xyz="0 0 0" rpy="${pi/2} 0 ${pi/2}"/>
      <mass value="${m}"/>
      <inertia ixx="${(m/12) * (h*h + d*d)}" ixy="0.0" ixz="0.0" iyy="${(m/12) * (w*w + d*d)}" iyz="0.0" izz="${(m/12) * (w*w + h*h)}"/>
    </inertial>
  </xacro:macro>

  <xacro:macro name="cylinder_inertia" params="m r h">
    <inertial>
      <origin xyz="0 0 0" rpy="${pi/2} 0 0" />
      <mass value="${m}"/>
      <inertia ixx="${(m/12) * (3*r*r + h*h)}" ixy = "0" ixz = "0" iyy="${(m/12) * (3*r*r + h*h)}" iyz = "0" izz="${(m/2) * (r*r)}"/>
    </inertial>
  </xacro:macro>

  <xacro:macro name="sphere_inertia" params="m r">
    <inertial>
      <mass value="${m}"/>
      <inertia ixx="${(2/5) * m * (r*r)}" ixy="0.0" ixz="0.0" iyy="${(2/5) * m * (r*r)}" iyz="0.0" izz="${(2/5) * m * (r*r)}"/>
    </inertial>
  </xacro:macro>
  

  
  
  <!-- Robot Base -->
  <link name="base_link"/>
  
  <xacro:macro name="octagon_face_box" params="name x y yaw">
    <link name="${name}_link">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <box size="${base_f_width} ${base_f_length} ${base_f_height}"/>
        </geometry>
        <material name="Cyan">
          <color rgba="0 1.0 1.0 1.0"/>
        </material>
      </visual>
      <collision>
        <geometry>
          <box size="${base_f_width} ${base_f_length} ${base_f_height}"/>
        </geometry>
      </collision>
      <xacro:box_inertia m="1.875" w="${base_f_width}" d="${base_f_length}" h="${base_f_height}"/>
    </link>
    
    <joint name="${name}_joint" type="fixed">
      <parent link="base_footprint"/>
      <child link="${name}_link"/>
      <origin xyz="${x} ${y} 0" rpy="0 0 ${yaw}"/>
    </joint>
  </xacro:macro>

  
  <xacro:octagon_face_box name="face_n" x="0.0" y="0.16" yaw="0"/>
  <xacro:octagon_face_box name="face_s" x="0.0" y="-0.16" yaw="0"/>
  <xacro:octagon_face_box name="face_e" x="0.16" y="0.0" yaw="${pi/2}"/>
  <xacro:octagon_face_box name="face_w" x="-0.16" y="0.0" yaw="${pi/2}"/>

  <xacro:macro name="octagon_corner_box" params="name x y yaw">
    <link name="${name}_link">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <box size="${base_e_width} ${base_e_length} ${base_e_height}"/>
        </geometry>
        <material name="Cyan">
          <color rgba="0 1.0 1.0 1.0"/>
        </material>
      </visual>
      <collision>
        <geometry>
          <box size="${base_e_width} ${base_e_length} ${base_e_height}"/>
        </geometry>
      </collision>
      <xacro:box_inertia m="1.875" w="${base_e_width}" d="${base_e_length}" h="${base_e_height}"/>
    </link>

    <joint name="${name}_joint" type="fixed">
      <parent link="base_footprint"/>
      <child link="${name}_link"/>
      <origin xyz="${x} ${y} 0" rpy="0 0 ${yaw}"/>
    </joint>
  </xacro:macro>

  
  <xacro:octagon_corner_box name="corner_ne" x="0.165" y="0.165" yaw="${pi/4}"/>
  <xacro:octagon_corner_box name="corner_nw" x="-0.165" y="0.165" yaw="${-pi/4}"/>
  <xacro:octagon_corner_box name="corner_se" x="0.165" y="-0.165" yaw="${-pi/4}"/>
  <xacro:octagon_corner_box name="corner_sw" x="-0.165" y="-0.165" yaw="${pi/4}"/>

  
  
  <!-- Robot Footprint -->
  <link name="base_footprint"/>  
  
  <joint name="base_joint" type="fixed">
    <parent link="base_footprint"/>
    <child link="base_link"/>
    <origin xyz="0.0 0.0 ${-(wheel_radius+wheel_zoff)}" rpy="0 0 0"/>
  </joint>
  
  <!-- Wheels -->
  <xacro:macro name="wheel" params="prefix x_reflect y_reflect">
    <link name="${prefix}_link">
      <visual>
        <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
        <geometry>
            <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
        </geometry>
        <material name="Gray">
          <color rgba="0.5 0.5 0.5 1.0"/>
        </material>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
        <geometry>
          <cylinder radius="${wheel_radius}" length="${wheel_width}"/>
        </geometry>
      </collision>

      <xacro:cylinder_inertia m="${wheel_mass}" r="${wheel_radius}" h="${wheel_width}"/>
    </link>

    <joint name="${prefix}_joint" type="continuous">
      <parent link="base_footprint"/>
      <child link="${prefix}_link"/>
      <origin xyz="${x_reflect * wheel_xoff} ${y_reflect * (wheel_yinset/2)} ${-wheel_zoff}" />
      <axis xyz="0 1 0"/>
      <limit effort="10.0" velocity="10.0"/>
    </joint>
  </xacro:macro>

  <xacro:wheel prefix="drivewhl_l" x_reflect="0" y_reflect="1" />
  <xacro:wheel prefix="drivewhl_r" x_reflect="0" y_reflect="-1" />
  
  <!-- Caster Wheel -->
  <xacro:macro name="caster_wheel" params="prefix x_pos y_pos">
    <link name="${prefix}_link">
      <visual>
        <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
        <geometry>
            <cylinder radius="${caster_radius}" length="${caster_width}"/>
        </geometry>
        <material name="Gray">
          <color rgba="0.5 0.5 0.5 1.0"/>
        </material>
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="${pi/2} 0 0"/>
        <geometry>
          <cylinder radius="${caster_radius}" length="${caster_width}"/>
        </geometry>
      </collision>

      <xacro:cylinder_inertia m="${caster_mass}" r="${caster_radius}" h="${caster_width}"/>
    </link>

  <joint name="${prefix}_joint" type="fixed">
    <parent link="base_footprint"/>
    <child link="${prefix}_link"/>
    <origin xyz="${x_pos} ${y_pos} ${-caster_zoff}" rpy="0 0 0"/>
  </joint>
  </xacro:macro>

  <!-- 앞쪽 좌우 -->
  <xacro:caster_wheel prefix="caster_fl" x_pos="${caster_xoff_front}" y_pos="${caster_ygap}" />
  <xacro:caster_wheel prefix="caster_fr" x_pos="${caster_xoff_front}" y_pos="${-caster_ygap}" />

  <!-- 뒤쪽 좌우 -->
  <xacro:caster_wheel prefix="caster_rl" x_pos="${caster_xoff_back}" y_pos="${caster_ygap}" />
  <xacro:caster_wheel prefix="caster_rr" x_pos="${caster_xoff_back}" y_pos="${-caster_ygap}" />
  
  
  <!-- 은판 -->
  <link name="silver_link">
    <visual>
      <geometry>
        <box size="${silver_length} ${silver_width} ${silver_height}"/>
      </geometry>
      <material name="gray">
        <color rgba="0.98 0.90 0.65 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="${silver_length} ${silver_width} ${silver_height}"/>
      </geometry>
    </collision>  
    <xacro:box_inertia m="0.5" w="${silver_width}" d="${silver_length}" h="${silver_height}"/>
  </link>
  
  <joint name="silver_joint" type="fixed">
    <parent link="base_footprint"/>
    <child link="silver_link"/>
    <origin xyz="0 0 ${base_height / 2 + silver_height / 2}" rpy="0 0 0"/>
  </joint>
  
    
  <!-- lidar 박스 -->
  <link name="box_link">
    <visual>
      <geometry>
        <box size="${box_length} ${box_width} ${box_height}"/>
      </geometry>
      <material name="gray">
        <color rgba="0.98 0.90 0.65 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <box size="${box_length} ${box_width} ${box_height}"/>
      </geometry>
    </collision>  
    <xacro:box_inertia m="0.5" w="${box_width}" d="${box_length}" h="${box_height}"/>
  </link>
  
  <joint name="box_joint" type="fixed">
    <parent link="base_footprint"/>
    <child link="box_link"/>
    <origin xyz="0 0 ${base_height / 2 + silver_height + box_height / 2}" rpy="0 0 0"/>
  </joint>

    
  
  <!-- lidar -->
  <link name="lidar_frame">
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <mass value="0.17"/>
      <inertia ixx="0.001"  ixy="0"  ixz="0" iyy="0.001" iyz="0" izz="0.002" />
      <!-- inertia 실제 값(아래) -->
      <!-- <inertia ixx="0.00024" iyy="0.00024" izz="0.00010" ixy="0" ixz="0" iyz="0" /> -->
    </inertial>

    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
         <cylinder radius="0.035" length="0.055"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
         <cylinder radius="0.035" length="0.055"/>
      </geometry>
      <material name="black">
        <color rgba="0 0 0 1"/>
      </material>
    </visual>
  </link>

  <joint name="lidar_joint" type="fixed">
    <parent link="box_link"/>
    <child link="lidar_frame"/>
    <origin xyz="0 0 ${box_height/2 + 0.0275}" rpy="0 0 0"/>
  </joint>
  
    <!-- 초음파 -->
  <xacro:macro name="ultrasonic_sensor" params="name parent_link x y z yaw topic">
    <link name="${name}_link">
      <visual>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <box size="${ultra_width} ${ultra_depth} ${ultra_height}"/>
        </geometry>
        <material name="Blue">
          <color rgba="0.2 0.5 0.8 1.0"/>
        </material>
      </visual>
      <collision>
        <geometry>
          <box size="${ultra_width} ${ultra_depth} ${ultra_height}"/>
        </geometry>
      </collision>
      <inertial>
        <mass value="${ultra_mass}"/>
        <inertia ixx="1e-6" iyy="1e-6" izz="1e-6" ixy="0" ixz="0" iyz="0"/>
      </inertial>
    </link>

    <joint name="${name}_joint" type="fixed">
      <parent link="${parent_link}"/>
      <child link="${name}_link"/>
      <origin xyz="${x} ${y} ${z}" rpy="0 0 ${yaw}"/>
    </joint>
  </xacro:macro>

  <xacro:ultrasonic_sensor name="echo_front" parent_link="base_footprint" x="0.325"  y="0.0"  z="0.085"  yaw="0.0"  topic="/echo_front"/> 
  <xacro:ultrasonic_sensor name="echo_back" parent_link="base_footprint" x="-0.325"  y="0.0"  z="0.085"  yaw="${pi}"  topic="/echo_back"/>
  <xacro:ultrasonic_sensor name="echo_right" parent_link="base_footprint" x="0.0"  y="-0.325"  z="0.085"  yaw="${-pi/2}"  topic="/echo_right"/>
  <xacro:ultrasonic_sensor name="echo_left" parent_link="base_footprint" x="0.0"  y="0.325"  z="0.085"  yaw="${pi/2}"  topic="/echo_left"/>
</robot> 
